sharding/syncer/handlers.go

```diff
@@ -0,0 +1,77 @@
+package syncer
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/mainchain"
+	"github.com/ethereum/go-ethereum/sharding/p2p"
+	"github.com/ethereum/go-ethereum/sharding/p2p/messages"
+)
+
+// RespondCollationBody is called by a node responding to another node's request
+// for a collation body given a (shardID, chunkRoot, period, proposerAddress) tuple.
+// The proposer will fetch the corresponding data from persistent storage (shardDB) by
+// constructing a collation header from the input and calculating its hash.
+func RespondCollationBody(req p2p.Message, signer mainchain.Signer, collationFetcher sharding.CollationFetcher) (*messages.CollationBodyResponse, error) {
+	// Type assertion helps us catch incorrect data requests.
+	msg, ok := req.Data.(messages.CollationBodyRequest)
+	if !ok {
+		return nil, fmt.Errorf("received incorrect data request type: %v", msg)
+	}
+
+	header := sharding.NewCollationHeader(msg.ShardID, msg.ChunkRoot, msg.Period, msg.Proposer, nil)
+	sig, err := signer.Sign(header.Hash())
```

Comment:
A proposer shouldn't have to sign again to reconstruct header to retrieve collation body for response. That means only the original proposer can respond with its collation body, and no one else can

The correct work flow should be:
1.  proposer creates, signs and adds header to smc
2. proposer broadcasts collation body to the rest of the network 
3. notary requests collation body to everyone in the network (not just the original proposer)
4. anyone can respond notary back with the collation body

couple ideas on how we can fix this;
1. we'll need to store proposer signature to SMC by adding a signature field to CollationRecord
2. we can just use collationBodyRequest's ChunkRoot to look up collationBody via `BodyByChunkRoot`. and we add `BodyByChunkRoot` method under interface `collationFetcher`

This just for future improvements and out of the scope within the given PR. I don't think we need to do this for Ruby release

Reply:
Great to know. You're right that this limits the nodes that can respond to requests. With this approach, observers cannot respond or store collation body information. Let's make a separate PR to add a proposer signature to SMC and then I can open a PR that fixes this.